### PR TITLE
Codebase review

### PR DESCRIPTION
**Pull request by vkomenda**
Part of the work for https://github.com/Fantom-foundation/libtransport-tcp/issues/1.

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
+    /// Indicating a vector reached max capacity and can not receive new element
     AtMaxVecCapacity,
     Bincode(bincode::Error),
     Io(std::io::Error),
-    // Indicating read/write operation was unable to read/write complete size of data
+    /// Indicating read/write operation was unable to read/write complete size of data
     Incomplete,
     PoisonError(String),
+    /// Socket address parse error.
+    AddrParse(AddrParseError),
```
Comment (Maxime2) [inline src/errors.rs]: Why this error enum has been added, I do not se any use of it in the code?

File: src/generic_test.rs [hunk 2]
```diff
@@ -25,6 +26,7 @@ impl From<usize> for Data {
     }
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, Hash, PartialEq, PartialOrd, Ord)]
```
Comment (Maxime2) [inline src/generic_test.rs]: TestPeer is not going to be serialized/deserialized, why these derived traits are added, they seems are not needed?

File: src/generic_test.rs [hunk 3]
```diff
@@ -34,14 +36,17 @@ impl Peer<Id> for TestPeer<Id> {
...
     fn get_id(&self) -> Id {
         self.id.clone()
     }
+
     fn get_net_addr(&self) -> String {
         self.net_addr.clone()
     }
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, Hash, PartialEq, PartialOrd, Ord)]
```
Comment (Maxime2) [inline src/generic_test.rs]: The same, why not used traits are added?

File: src/lib.rs [hunk 4]
```diff
@@ -5,52 +5,42 @@ use serde::de::DeserializeOwned;
...
 
-// Transport configurtatiion trait
+/// Transport configuration trait.
 pub trait TransportConfiguration<Data> {
-    // creates new transport configuration with specified network
-    // address for incoming messages listener
-    fn new(set_bind_net_addr: String) -> Self;
+    /// Creates a new transport configuration with the specified network address for incoming
+    /// message listener.
+    fn new(set_bind_net_addr: &str) -> Result<Box<Self>>;
```
Comment (Maxime2) [inline src/lib.rs]: What is the reson to change result type to `Result<Box<Self>>` ?

File: src/lib.rs [hunk 5]
```diff
@@ -5,52 +5,42 @@ use serde::de::DeserializeOwned;
...
 
-    // set bind network address for incoming messages listener
-    fn set_bind_net_addr(&mut self, address: String) -> Result<()>;
+    /// Sets the bind network address of the incoming message listener.
+    fn set_bind_net_addr(&mut self, address: &str) -> Result<()>;
```
Comment (Maxime2) [inline src/lib.rs]: What are reason to change type if address from `String` to `&str`?

File: src/errors.rs (hunk 1 above)
Comment (vkomenda) [reply src/errors.rs]: Here: https://github.com/Fantom-foundation/libtransport-tcp/blob/0cec040bfe556f33d5d4bb04f3501280f792a9c8/src/lib.rs#L31

File: src/generic_test.rs (hunk 2 above)
Comment (vkomenda) [reply src/generic_test.rs]: If the instances are derivable, then it is advisable to define them following the [API convention](https://rust-lang-nursery.github.io/api-guidelines/interoperability.html#c-common-traits).

File: src/generic_test.rs (hunk 3 above)
Comment (vkomenda) [reply src/generic_test.rs]: Same as above.

File: src/lib.rs (hunk 4 above)
Comment (vkomenda) [reply src/lib.rs]: Because `Self` is a trait object which is not `Sized` and therefore cannot be passed by value inside the `Result`, only by a fat pointer.

You do need to return a `Result` and any possible errors. It's not an option to `panic` inside those methods.

I generally recommend to avoid trait objects because they lead to much slower running code compared to static dispatch.

File: src/lib.rs (hunk 5 above)
Comment (vkomenda) [reply src/lib.rs]: I provided an explanation [here](https://github.com/Fantom-foundation/libtransport-tcp/pull/3#issuecomment-525210387).

File: src/generic_test.rs (hunk 2 above)
Comment (Maxime2) [reply src/generic_test.rs]: It is not API at all!

File: Cargo.toml [hunk 6]
```diff
@@ -1,12 +1,16 @@
...
+]
 edition = "2018"
 
 [dependencies]
 libcommon-rs = { git = "https://github.com/Fantom-foundation/libcommon-rs" }
 serde = "1.0.97"
 bincode = "1.0.1"
 os_pipe = "0.8.0"
-futures-preview = { version = "=0.3.0-alpha.17", features = ["async-await", "nightly"] }
+futures-preview = { version = "=0.3.0-alpha.18", features = ["async-await", "nightly"] }
```
Comment (Maxime2) [inline Cargo.toml]: Version change for futures-preview should be changed in all repos altogether to avoid version conflicts, so it is better to do in separate PR and having a reason to do so due to use features unavailable in current version, which is seem is not the case for this PR.

File: src/errors.rs (hunk 1 above)
Comment (Maxime2) [reply src/errors.rs]: That is not necessary as bind() would return io::Error if unable to parse address from string.

File: src/lib.rs (hunk 4 above)
Comment (Maxime2) [reply src/lib.rs]: You can add `Sized` bound to `Self, no need to put it into `Box`.

File: Cargo.toml (hunk 6 above)
Comment (vkomenda) [reply Cargo.toml]: OK. This is not really important. We can keep the version.

File: src/lib.rs (hunk 4 above)
Comment (vkomenda) [reply src/lib.rs]: No. Sizes of objects of instances of `Transport` are not known at compile time.

File: src/generic_test.rs (hunk 2 above)
Comment (vkomenda) [reply src/generic_test.rs]: It is a public type in a library. So it is possible of use it outside the library. So technically it **is** part of the library interface. You mat want to change the visibility level if you don't intended that.

File: src/errors.rs (hunk 1 above)
Comment (vkomenda) [reply src/errors.rs]: `AddrParseError` is the error returned by the parser. Let's not drop errors when they are given.

File: src/lib.rs (hunk 4 above)
Comment (vkomenda) [reply src/lib.rs]: @Maxime2 I see you added the annotation [yourself](https://github.com/Fantom-foundation/libtransport/blob/c93dd87fb57232738018932c699dbb8535df5baf/src/lib.rs#L19). It's not going to help because the compiler cannot determine the size of `TcpTransportCfg`. So could you please remove it.

Comment (SamuelMarks) [reply src/lib.rs]: Agreeing with @vkomenda here. We've got to be extra careful when writing Rust code to handle paths where sizes are unknown until runtime.

**Comment by vkomenda** [comment]
@SamuelMarks, @Maxime2, please have a look. I fixed a few more irregularities.

File: src/lib.rs (hunk 4 above)
Comment (vkomenda) [reply src/lib.rs]: I decided to require `Sized` for `TransportConfiguration` and have no such requirement in `Transport`. That hasn't presented any problem yet. If it does, we can go back to boxing the returned value of `TransportConfiguration::new` and this is not going to introduce overhead because configuration steps are infrequent.

File: src/errors.rs (hunk 1 above)
Comment (Maxime2) [reply src/errors.rs]: Bind() function uses https://github.com/rust-lang/rust/blob/09189591c4c4f6784ffd4bbe99eaefbfe1d5e4a4/src/libstd/net/addr.rs#L906 (see https://github.com/rust-lang/rust/blob/c84a7abf8b2c5753179472464dc2baeb86c6fed6/src/libstd/net/tcp.rs#L678 as well) to convert from string to socket address and returns io::Error, so this change is not needed as io::Error is already covered.

File: src/generic_test.rs (hunk 2 above)
Comment (Maxime2) [reply src/generic_test.rs]: It is not intended to be used outside library. If you can make it private, you're welcome with another PR. Otherwise these derived traits are not needed.

File: src/generic_test.rs [hunk 7]
```diff
@@ -101,31 +107,34 @@ impl PeerList<Id, Error> for TestPeerList<Id> {
...
-    THis method simply takes in a list of peers, instantiates them, and tests whether they can
-    send/receive data to one another.
-*/
+    This method simply takes in a list of peers, instantiates them, and tests whether they can
+    send/receive data to/from one another.
+**/
 pub fn common_test<
-    //    C: TransportConfiguration<Data>,
-    T: Transport<Id, Data, Error, TestPeerList<Id>>,
+    C: TransportConfiguration<Data>,
```
Comment (Maxime2) [inline src/generic_test.rs]: That was removed in the latest master branch, why it is reverted back here?

**Comment by Maxime2** [comment]
Why do you deleting newly added comments to the code without any explanation why you do so?

**Comment by vkomenda** [comment]
@Maxime2, we have to reach consensus on the review and merge the result.

File: src/errors.rs (hunk 1 above)
Comment (vkomenda) [reply src/errors.rs]: OK, I can leave the string argument and the `bind` function and remove this error. My idea was that having an address of type `SocketAddr` would have add static correctness. But it's not important.

File: src/generic_test.rs (hunk 2 above)
Comment (vkomenda) [reply src/generic_test.rs]: I'll make that in this PR.

File: src/generic_test.rs (hunk 7 above)
Comment (vkomenda) [reply src/generic_test.rs]: Because that change in `master` is a regression and I did not agree with it. It is better to pass a configuration object to `Transport::new`.

**Comment by Maxime2** [comment]
@vkomenda you were tasked to review existing code, i.e. make comments to it, not to improve it or correct documentation for it as a PR. It was expected to be done overnight here, i.e. on your next business day, you have delivered it several days later than expected, with changes not wanted and now you're insisting on your changes to be merged after they were rejected and reverting changes made after your PR created. That is irrational and quite obnoxious for me.

Please make changes to documentation you see needed to be done in separate PR against current master, and changes in the [commment](https://github.com/Fantom-foundation/libtransport/pull/2#discussion_r320074238) in another separate PR against current master.
